creusot-contracts/src/ghost/ptr_own.rs

```diff
@@ -235,16 +210,316 @@ impl<T: ?Sized> PtrOwn<T> {
+    pub fn as_ptr_mut(&mut self) -> &mut PtrMut<'_, T> {
+        panic!("called ghost function in normal code")
+    }
+}
+
+/// # Permissions for slice pointers
+///
+/// Core methods:
+///
+/// - To split a `&PtrOwn<[T]>`: [`split_at`](PtrOwn::split_at).
+/// - To index a `&PtrOwn<[T]>` into `&PtrOwn<T>`: [`elements`](PtrOwn::elements).
+/// - To extract a [`&PtrLive<T>`][PtrLive] (evidence used by pointer arithmetic): [`live`](PtrOwn::live).
+impl<T> PtrOwn<[T]> {
+    /// The number of elements in the slice.
+    #[logic(open, inline)]
+    pub fn len(self) -> Int {
+        pearlite! { self.val()@.len() }
+    }
+
+    /// Split a `&PtrOwn<[T]>` into two subslices of lengths `index` and `self.len() - index`.
+    #[trusted]
+    #[check(ghost)]
+    #[requires(0 <= index && index <= self.len())]
+    #[ensures(self.ptr().thin() == result.0.ptr().thin())]
+    #[ensures(self.ptr().thin().offset_logic(index) == result.1.ptr().thin())]
+    #[ensures(self.val()@.subsequence(0, index) == result.0.val()@)]
+    #[ensures(self.val()@.subsequence(index, self.len()) == result.1.val()@)]
+    pub fn split_at(&self, index: Int) -> (&Self, &Self) {
+        let _ = index;
+        panic!("called ghost function in normal code")
+    }
+
+    /// Split `&PtrOwn<[T]>` into a sequence of `&PtrOwn<T>` for each element.
+    #[trusted]
+    #[check(ghost)]
+    #[ensures(result.len() == self.len())]
+    #[ensures(forall<i> 0 <= i && i < self.len()
+        ==> result[i].ptr() == self.ptr().thin().offset_logic(i)
+        && *result[i].val() == self.val()@[i])]
+    pub fn elements(&self) -> Seq<&PtrOwn<T>> {
+        panic!("called ghost function in normal code")
+    }
+
+    /// Convert a `&PtrOwn<[T]>` for a non-empty slice into a `&PtrOwn<T>` for the element at the given index.
+    #[check(ghost)]
+    #[requires(0 <= index && index < self.len())]
+    #[ensures(result.ptr() == self.ptr().thin().offset_logic(index))]
+    #[ensures(*result.val() == self.val()@[index])]
+    pub fn index(&self, index: Int) -> &PtrOwn<T> {
+        let mut r = self.elements();
+        r.split_off_ghost(index).pop_front_ghost().unwrap()
+    }
+
+    /// Extract `PtrLive<'a, T>` from `&'a PtrOwn<[T]>`
+    #[trusted]
+    #[check(ghost)]
+    #[ensures(result.ptr() == self.ptr().thin())]
+    #[ensures(result.len()@ == self.len())]
+    pub fn live(&self) -> PtrLive<'_, T> {
+        panic!("called ghost function in normal code")
+    }
+}
+
+/// Permission token for mutable access to a pointer.
+///
+/// This is used by functions that return mutable borrows of tokens (e.g., `from_mut`, `split_at_mut`, `elements_mut`).
+/// If we returned `&mut PtrOwn<T>`, that would allow the caller to move the `PtrOwn<T>` out of the borrow
+/// and keep it even after the end of the borrow.
+/// Instead, we return `&'a mut PtrMut<'a, T>` so that the token cannot outlive the borrow.
+#[allow(dead_code)]
+#[opaque]
+pub struct PtrMut<'a, T: ?Sized>(PhantomData<&'a T>);
```

Comment:
I don't like the idea of having two different tokens here.

Could we make `PtrOwn<T>` an alias for `PtrMut<'static, T>`?